TYP: improved ``numpy.array`` type hints for array-like input

guan404ming opened the pull request:
This partially fixes https://github.com/numpy/numpy/issues/25691 by adding an overload to ``numpy.array`` for array-likes input of numpy scalar types.

jorenham commented:
For most of the input types, this overlaps with https://github.com/numpy/numpy/pull/26945, but neither one is technically speaking better than the other:
Whereas this PR also takes the nested sequences of builtin (scalar) types into account, 
https://github.com/numpy/numpy/pull/26945 covers more scalar types, e.g. array-likes with `object_` dtype.

The `_ArrayLike{}_co` overload approach that you've used here, is a rather common one within the numpy stubs.
But I recently found out that this approach comes with a rather tricky problem, and is caused by overlapping overloads with incompatible return types.

But a piece of code speaks more than a thousand words:

```python
import numpy as np

list_int: list[int]
list_float: list[float]
list_int_float: list[int | float]
list_int_list_float: list[int] | list[float]

reveal_type(np.array(list_int))
reveal_type(np.array(list_float))
reveal_type(np.array(list_int_float))
reveal_type(np.array(list_int_list_float))
```

Using this PR's branch, and the code above (as `.pyi`), Pyright reports the following:

1. `Type of "np.array(list_int)" is "ndarray[Any, dtype[signedinteger[Any]]]"`
2. `Type of "np.array(list_float)" is "ndarray[Any, dtype[floating[Any]]]"`
3. `Type of "np.array(list_int_float)" is "ndarray[Any, dtype[floating[Any]]]"`
4. `Type of "np.array(list_int_list_float)" is "ndarray[Any, dtype[floating[Any]]]"`

But 3. and 4. aren't correct: their scalar types should be the union of 1. and 2., i.e. `signedinteger[Any] | floating[Any]`.

The overlap in overloads is tricky to deal with, but cannot be fully solved using the `_ArrayLike{}_co` aliases.
I believe that the best solution here would lie somewhere between your approach and mine, and requires separate overloads for the builtin python types and the numpy-scalar types.

In any case, I'd recommend adding type-tests to go along with your change, including ones that test for this overlapping-overload issue.

guan404ming commented:
Hi @jorenham, I read your feedback and it makes a lot of sense. Adding a suite of testing is definitely a good idea, and I’ll also consider your suggestions to improve this PR. Thanks a lot for your suggestions. Maybe we can work together to fix this issue.

jorenham commented:
There's some more information about this "overlapping-overloads " issue in https://github.com/numpy/numpy/issues/27032

jorenham reviewed (changes requested):
See https://github.com/numpy/numpy/pull/27019#issuecomment-2246451848

guan404ming commented:
@jorenham I've gone ahead and added some tests as you suggested. I'm still not entirely sure which approach is best here, so I was wondering if you could provide any further guidance on how to improve this? Maybe you could share some example code that shows a better way to handle these tricky cases? I'd really appreciate any suggestions you might have. Thanks again for your help!

jorenham commented:
> I'm still not entirely sure which approach is best here, so I was wondering if you could provide any further guidance on how to improve this? Maybe you could share some example code that shows a better way to handle these tricky cases?

The solution to this problem can be separated into two parts: 

1. The numpy part: Array-like that's expressible in terms of (subtypes of) `numpy.generic`
2. The python part: These are the (nested) sequences that contain python builtin scalars, e.g. `[[42]]`.

The numpy part is the easy one, because numpy's scalar types don't overlap.
I'd suggest using `numpy._typing._ArrayLike` for this, so that you can do this in a single overload, similar to https://github.com/numpy/numpy/pull/26945/files#diff-d8e4a975af9665843eeb7d5d1670b835064d8fdfa37ccfb6c2640a296085d54fR448-R458 .

The python part is the tricky one, because of the overlap issue where e.g. `x: float` also accepts `int` and `bool`.
These will require separate overloads for nested sequences (`numpy._typing._NestedSequence`) of `bool`, `int`, `float`, `complex`, and `object` (in that order).
But for each overload, you're gonna have to make sure that the return types are compatible with the previous, overlapping, overloads. So that the overload for e.g. `float` input as a return type that includes not only `np.floating[Any]`, but also `np.integer[Any]` and `np.bool`.

Alternatively, you could use a `Protocol`-based approach, so that you don't have overlap anymore, e.g. with an `OnlyFloat` protocol that describes "a float that's not an int".
In https://github.com/numpy/numpy/issues/27032#issuecomment-2249308801 I posted a minimal example of this that works in Pyright, but requires more tests to verify if it works with mypy, and tests to check if it correctly deals with the numeric numpy scalar types.

> I'd really appreciate any suggestions you might have. Thanks again for your help!

And thank you for you interest in numpy typing :)

guan404ming commented:
I've made good progress on the numpy part and have successfully completed the tests. However, I'm still facing some challenges with the python part, particularly with the overlapping issues. I attempted to implement your suggestion using `Protocols`, but I'm encountering some difficulties with `int` and `bool`. I'm not sure if it's due to an error in my setup, but I'll continue to investigate.

Given that these overlapping issues are somewhat separate from the main focus of this PR and involve a broader scope, I was wondering if it might be appropriate to merge this PR as it stands now. We could then continue the discussion about the overlapping problems in more depth over at #27032. My concern is that addressing these issues here might make this PR overly complex.

Of course, I'm open to your thoughts on this. If you have any other suggestions or if you think I should make any change. Thank you again for all your assistance with numpy typing.

jorenham commented:
> Given that these overlapping issues are somewhat separate from the main focus of this PR and involve a broader scope, I was wondering if it might be appropriate to merge this PR as it stands now. 

I agree; it would be better if we keep those Protocols out of this change, and stick to `np.array`.

jorenham commented on numpy/_core/multiarray.pyi:
```diff
@@ -199,6 +218,72 @@ def array(
+    *,
+    copy: None | bool | _CopyMode = ...,
+    order: _OrderKACF = ...,
+    subok: bool = ...,
+    ndmin: int = ...,
+    like: None | _SupportsArrayFunc = ...,
+) -> NDArray[signedinteger[Any]]: ...
+@overload
+def array(
+    object: _ArrayLikeFloat_co,
+    dtype: None = ...,
+    *,
+    copy: None | bool | _CopyMode = ...,
+    order: _OrderKACF = ...,
+    subok: bool = ...,
+    ndmin: int = ...,
+    like: None | _SupportsArrayFunc = ...,
+) -> NDArray[floating[Any]]: ...
+@overload
+def array(
+    object: _ArrayLikeComplex_co,
+    dtype: None = ...,
+    *,
+    copy: None | bool | _CopyMode = ...,
+    order: _OrderKACF = ...,
+    subok: bool = ...,
+    ndmin: int = ...,
+    like: None | _SupportsArrayFunc = ...,
+) -> NDArray[complexfloating[Any, Any]]: ...
+@overload
+def array(
+    object: _ArrayLikeTD64_co,
+    dtype: None = ...,
+    *,
+    copy: None | bool | _CopyMode = ...,
+    order: _OrderKACF = ...,
+    subok: bool = ...,
+    ndmin: int = ...,
+    like: None | _SupportsArrayFunc = ...,
+) -> NDArray[timedelta64]: ...
```
These `_ArrayLike{}_co` aliases should be replaced with the `T | _NestedSequence[T]` part from [their definition](https://github.com/guan404ming/numpy/blob/55c10be2ebb451b80924016a3628866c857827b8/numpy/_typing/_array_like.py#L96-L128), because they also include `_ArrayLike`, which you already covered.

So e.g. `_ArrayLikeFloat_co` will become `float | _NestedSequence[float]` (the `int` and `bool` are implied by `float`, because the type param of `_NestedSequence` is covariant). 
And because `float` could also come as a union with `int` or `bool`, the return type should be changed to `NDArray[np.bool | np.int_ | np.float64]` (note the narrowed `floating` and `integer`), so that it's always compatible with the (previous) overload for `int | _NestedSequence[int]`.

jorenham commented on numpy/typing/tests/data/reveal/array_constructors.pyi:
```diff
@@ -37,6 +40,10 @@ assert_type(np.array(deque([1, 2, 3])), npt.NDArray[Any])
 assert_type(np.array(A, like=A), npt.NDArray[np.float64])
+assert_type(np.array(D), npt.NDArray[np.float64 | np.int64])
+assert_type(np.array(C), npt.NDArray[np.signedinteger[Any]])
+assert_type(np.array(E), npt.NDArray[np.floating[Any]])
```
`list[float]` can also contain `int` or `bool`, so we want the returned array to have a scalar type of `float64 | np.int_ | np.bool`.

jorenham commented on numpy/typing/tests/data/reveal/array_constructors.pyi:
```diff
@@ -20,6 +20,9 @@ i8: np.int64
 C: list[int]
+D: SubClass[np.float64 | np.int64]
+E: list[float]
+F: list[bool]
```
to fully check whether the overlapping overloads are all compatible, you could add an explicit test for something like `list[float] | list[int]`

jorenham replied:
what will happen with `list[str]`, `list[bytes]`, or `list[decimal.Decimal]` 🤔?

jorenham commented on numpy/typing/tests/data/reveal/array_constructors.pyi:
```diff
@@ -37,6 +40,10 @@ assert_type(np.array(deque([1, 2, 3])), npt.NDArray[Any])
+assert_type(np.array(D), npt.NDArray[np.float64 | np.int64])
+assert_type(np.array(C), npt.NDArray[np.signedinteger[Any]])
+assert_type(np.array(E), npt.NDArray[np.floating[Any]])
+assert_type(np.array(F), npt.NDArray[np.bool])
```
I'm missing the tests for `ndarray` subtypes and the possible `subok` / `ndmin` options, so that the first two overloads are properly checked.

jorenham commented on numpy/_core/multiarray.pyi:
```diff
@@ -188,6 +196,17 @@ def array(
     like: None | _SupportsArrayFunc = ...,
 ) -> _ArrayType: ...
```
Because the bounds of `_ArrayType` are now changed so that they also include a potential shape type, the `ndmin` parameter is now able to change the return type when `ndmin > 0`.

jorenham commented on numpy/typing/tests/data/reveal/array_constructors.pyi:
```diff
@@ -20,6 +20,9 @@ i8: np.int64
 B: SubClass[np.float64]
 C: list[int]
+D: SubClass[np.float64 | np.int64]
+E: list[float]
```
a simple test for a bare scalar could be nice, e.g. `np.array(3.14)`

jorenham commented on numpy/typing/tests/data/reveal/array_constructors.pyi:
```diff
@@ -20,6 +20,9 @@ i8: np.int64
 A: npt.NDArray[np.float64]
 B: SubClass[np.float64]
 C: list[int]
+D: SubClass[np.float64 | np.int64]
```
will this also work for custom type that implements `__array__`, but isn't a subtype of `ndarray`?

guan404ming commented:
Thank you for your suggestions on the type definitions. I've implemented the changes as proposed:

1. **Replaced ArrayLike{}_co aliases with (T | _NestedSequence[T])**, simplifying our code significantly.

2. **Kept the return type unchanged.** Changing it to NDArray[np.bool | np.int_ | np.float64] could potentially cause major disruptions to existing code and dependencies. This would require a more extensive discussion and careful planning before implementation.

3. **Implemented separate handling for bytes and str types, and added specific overloads for list[str] and list[bytes].** These changes are reflected in the updated tests.

These improvements enhance clarity without breaking compatibility. I've tested the implementation. Please let me know if you'd like to review or discuss further, especially regarding potential future changes to the return type.

jorenham reviewed (changes requested):
I realize that these comments about the overlapping overloads can come across as nitpicking, since this overlapping-overloads issue is a very common one in numpy.

But look at it this way:
As it currently stands, the type annotations of `array` are fully correct, as it accept everything that is also allowed at runtime, and in each case its return type is correct, even though often  unnecessarily broad (i.e. `NDArray[Any]`).
So your goal is to narrow the return types, which you've already accomplished for the majority of the cases.
But at the moment, your approach still has overlapping-overloads with incompatible return types. If I were to accept it as it is now, this would introduce cases for which the `np.array` types are incorrect, which previously wasn't the case.
And I'd be the first to admit that it's very weird that `np.array(42)` should be typed to return `NDArray[np.int_ | np.bool]`, but unfortunately, that is the only way that's *always* correct.
Perhaps I'm repeating myself, but here's another example that shows why this is necessary:

```python
def scalar2array(x: int) -> ???:
    return np.array(x)
```

At runtime `scalar2array(42)` returns an `np.int_` array, and calling `scalar2array(True)` returns a `np.bool` array.
But what if you pass something that can be *either* `42` or `True`, e.g. `random.choice((42, True))`? Then the result should have a `np.int_ | np.bool` dtype, right?
Well, the way that it's currently types, you'd get just `np.int_`,
because after all, `int` and `int | bool` are indistinguishable for type-checkers.

So if we were to leave these incompatible overlapping-overloads in, then we'd introduce scenario's like in this example, where the return type incorrect. And I'm sure you can see how this can be very confusing.

And unfortunately, in the typing specs it was decided that, contrary to runtime, type-checkers should consider also accept an `int` in place of a `float`, and a `float` in place of a `complex`. So the only way to deal with it, is with those `Only*` protocols, which we've decided to leave out of this PR.
So the only type-correct option that remains, is for us to annotate `scalar2array` to have a return type of `NDArray[np.int_ | np.bool]`, as weird as it may look.

And to help onlookers understand, it could help if we'd place something like `# see gh issue #27032` near it.

---

Edit:

If you're not comfortable implementing these overlapping-overload changes, then that's no problem either. You could just leave those overloads out of it, so that in those cases, the behaviour doesn't change (i.e. returns `NDArray[Any]` as it currently does).
The remaining added overload for `_ArrayLike[_SCT]` will already be an improvement, and doesn't introduce any issues on its own I believe.

guan404ming commented:
@jorenham Thank you for your feedback. After considering your points, I still think that the overlapping overloads require further discussion. For now, I keep the remaining added overload for _ArrayLike[_SCT] and modify its return type to NDArray[Any]. This should allow Pyright to correctly infer the types, which I believe is an improvement. I've also removed some redundant tests that no longer apply after these changes.

jorenham commented on numpy/_core/multiarray.pyi:
```diff
@@ -199,6 +218,72 @@ def array(
+    *,
+    copy: None | bool | _CopyMode = ...,
+    order: _OrderKACF = ...,
+    subok: bool = ...,
+    ndmin: int = ...,
+    like: None | _SupportsArrayFunc = ...,
+) ->NDArray[Any]: ...
+@overload
+def array(
+    object: int | _NestedSequence[int],
+    dtype: None = ...,
+    *,
+    copy: None | bool | _CopyMode = ...,
+    order: _OrderKACF = ...,
+    subok: bool = ...,
+    ndmin: int = ...,
+    like: None | _SupportsArrayFunc = ...,
+) -> NDArray[Any]: ...
+@overload
+def array(
+    object: float | _NestedSequence[float],
+    dtype: None = ...,
+    *,
+    copy: None | bool | _CopyMode = ...,
+    order: _OrderKACF = ...,
+    subok: bool = ...,
+    ndmin: int = ...,
+    like: None | _SupportsArrayFunc = ...,
+) -> NDArray[Any]: ...
+@overload
+def array(
+    object: complex | _NestedSequence[complex],
+    dtype: None = ...,
+    *,
+    copy: None | bool | _CopyMode = ...,
+    order: _OrderKACF = ...,
+    subok: bool = ...,
+    ndmin: int = ...,
+    like: None | _SupportsArrayFunc = ...,
+) -> NDArray[Any]: ...
```
These ones are not needed, as they are already covered by the last overload

jorenham commented on numpy/typing/tests/test_typing.py:
```diff
@@ -20,10 +20,10 @@
-    not RUN_MYPY,
-    reason="`NPY_RUN_MYPY_IN_TESTSUITE` not set"
-)
+# pytestmark = pytest.mark.skipif(
+#     not RUN_MYPY,
+#     reason="`NPY_RUN_MYPY_IN_TESTSUITE` not set"
+# )
```
this doesn't seem like it was intentional 🤔

jorenham commented on numpy/typing/tests/data/reveal/array_constructors.pyi:
```diff
@@ -38,6 +39,21 @@ assert_type(np.array(A, dtype=np.int64), npt.NDArray[np.int64])
+assert_type(np.array(A, subok=True), npt.NDArray[np.float64])
+assert_type(np.array(D), npt.NDArray[np.float64 | np.int64])
+
+assert_type(np.array(True), npt.NDArray[Any])
+assert_type(np.array(-1), npt.NDArray[Any])
+assert_type(np.array(3.14), npt.NDArray[Any])
+assert_type(np.array(b"qwe"), npt.NDArray[Any])
+assert_type(np.array("qwe"), npt.NDArray[Any])
+assert_type(np.array([1, 5, 6]), npt.NDArray[Any])
+assert_type(np.array([1.0, 5.1, 6]), npt.NDArray[Any])
+assert_type(np.array(["qwe", 5, 6]), npt.NDArray[Any])
+
```
These tests are for the overloads that weren't added, so they can be removed

```suggestion
```

jorenham commented on numpy/typing/tests/data/reveal/array_constructors.pyi:
```diff
@@ -38,6 +39,21 @@ assert_type(np.array(A, dtype=np.int64), npt.NDArray[np.int64])
 assert_type(np.array(A, like=A), npt.NDArray[np.float64])
 
+assert_type(np.array(B, subok=True), SubClass[np.float64])
+assert_type(np.array(B, subok=True, ndmin=0), SubClass[np.float64])
```
passing e.g. `ndmin=1` should result in different output type

jorenham reviewed (changes requested):
I left some minor suggestions.
After addressing those, and cleanup up the commit history (see https://numpy.org/doc/stable/dev/development_workflow.html#rewriting-commit-history), then I think that we can merge this

guan404ming committed:
TYP: enhance array type

jorenham commented:
Thanks @guan404ming !

guan404ming commented:
Thank you, @jorenham ! I really appreciate your guidance and assistance on this pr. I've learned a great deal from your expertise~
